draft-ietf-quic-recovery.md

```diff
@@ -601,7 +601,9 @@ removed from bytes in flight when the Initial and Handshake keys are discarded.
 When a PTO timer expires, a sender MUST send at least one ack-eliciting packet
 as a probe, unless there is no data available to send.  An endpoint MAY send up
 to two full-sized datagrams containing ack-eliciting packets, to avoid an
-expensive consecutive PTO expiration due to a single lost datagram.
+expensive consecutive PTO expiration due to a single lost datagram.  When only
+sending a single packet on PTO, senders can skip a packet number to elicit a
+faster acknowledgement.
 
```

Comment:
I don't understand this: while packet numbers are usually incremental, that does not carry any semantics to my knowledge. PN's are skipped to guard against optimistic ACK attacks. It would be unfortunate if that praxis affects retransmission?

That said, I think it would be great if PN's could be assumed to be incremental, but there is no good solution to the optimistic ACK attack so far.

Reply:
IIUC, skipping a packet number is a way to elicit an immediate ACK.

Quoting from section 4: _In order to accelerate loss recovery and reduce timeouts, the receiver SHOULD send an immediate ACK after it receives an out-of-order packet._

Reply:
But how is "out-of-order" defined?

Reply:
So if both endpoints skip every second packet number early in the connection, in order to aggressively detect optimistic ACK's, would that lead to an ACK flood?

Reply:
The term is defined in [section 2](https://quicwg.org/base-drafts/draft-ietf-quic-recovery.html#rfc.section.2).

Reply:
Why are we recommending to skip packet numbers over just including a PING? The former causes more overhead in the reverse direction.

Reply:
@larseggert IIUC, the receiver of a PING is not required to send an ACK immediately.

My slight preference goes to handling the issue below the framing layer, because it'd be only after you've generated the payload of a packet (by gathering frames) that you discover that the packet that is going to be sent is a tail.